common/src/main/java/com/hortonworks/registries/common/util/HdfsFileStorage.java

```diff
@@ -33,68 +44,155 @@
+            kerberos = true;
         }
 
-        try {
-            hdfsFileSystem = FileSystem.get(URI.create(fsUrl), hdfsConfig);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+        directory = adjustDirectory(fsUrl, directory);
+        fsUri = URI.create(fsUrl);
+
+        LOG.info("Initialized with fsUrl={}, directory={}, kerberos principal={}", fsUrl, directory, kerberosPrincipal);
+    }
+
+    /**
+     * Cloud storage filesystem url's usally contain paths. These should be prepended to the directory in order to work
+     * properly.
+     * @param fsUrl The HDFS or compatible filesystem URL
+     * @param directory the directory to store hars
+     * @return directory adjusted with path component in fsUrl
+     */
+    @VisibleForTesting
+    static String adjustDirectory(String fsUrl, String directory) {
+        String pathInFsUrl = URI.create(fsUrl).getPath();
+        if ( !(pathInFsUrl.endsWith("/") || directory.startsWith("/")) ) {
+            return pathInFsUrl + "/" + directory;
+        }
+        else if (pathInFsUrl.endsWith("/") && directory.startsWith("/")) {
+            return pathInFsUrl + directory.substring(1);
+        }
+        else {
+            return pathInFsUrl + directory;
         }
     }
 
+    private boolean isKerberos() {
+        return kerberos;
+    }
+
     @Override
     public String upload(InputStream inputStream, String name) throws IOException {
-        Path jarPath = new Path(directory, name);
-
-        try(FSDataOutputStream outputStream = hdfsFileSystem.create(jarPath, false)) {
-            ByteStreams.copy(inputStream, outputStream);
-        }
-
-        return jarPath.toString();
+        return isKerberos() ?
+                doAsAndConvertException(() -> uploadInternal(inputStream, name)):
+                uploadInternal(inputStream, name);
     }
 
     @Override
     public InputStream download(String name) throws IOException {
-        Path filePath = new Path(directory, name);
-        return hdfsFileSystem.open(filePath);
+        return isKerberos() ?
+                doAsAndConvertException(() -> downloadInternal(name)):
+                downloadInternal(name);
     }
 
     @Override
     public boolean delete(String name) throws IOException {
-        return hdfsFileSystem.delete(new Path(directory, name), true);
+        return isKerberos() ?
+                doAsAndConvertException(() -> deleteInternal(name)):
+                deleteInternal(name);
     }
 
     @Override
     public boolean exists(String name) {
         try {
-            return hdfsFileSystem.exists(new Path(directory, name));
+            return isKerberos() ?
+                    UserGroupInformation.getLoginUser().doAs((PrivilegedAction<Boolean>) () -> existsInternal(name)):
+                    existsInternal(name);
+        }
+        catch (IOException ex) {
+            throw new UncheckedIOException(ex);
```

Comment:
Why are we wrapping the IOException as UncheckedIOException here but not in other internal methods?

Reply:
This method didn't declare IOException unlike the others (I don't know the reason for this). Adding FileSystem initialization brought in IOException, but I didn't want to change the signature as FS initialization may be temporary (until IDbroker issues are fixed).

I think the correct solution for this is that this method would declare IOException just like others and it wouldn't swallow exceptions (It currently swallows IOException in existsInternal()). How about that?

Reply:
Make sense. We should let exists() throw IOException.